requests/utils.py

```diff
@@ -87,7 +83,23 @@ def super_len(o):
             # is actually a special file descriptor like stdin. In this
             # instance, we don't know what the length is, so set it to zero and
             # let requests chunk it instead.
-            current_position = total_length
+            if total_length is not None:
+                current_position = total_length
+
+    if hasattr(o, 'seek') and total_length is None:
+        # StringIO and BytesIO have seek but no useable fileno
+
+        # seek to end of file
+        o.seek(0, 2)
+        if hasattr(o, 'tell'):
```

Comment:
Yeah, good spot. This block does nothing sensible if there is no `tell`, so we need to check for both.